vm/compiler/src/coinbase_puzzle/helpers/coinbase_solution/serialize.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+impl<N: Network> Serialize for CoinbaseSolution<N> {
+    /// Serializes the coinbase solution to a JSON-string or buffer.
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        match serializer.is_human_readable() {
+            true => {
+                let mut combined_puzzle_solution = serializer.serialize_struct("CoinbaseSolution", 3)?;
+                combined_puzzle_solution.serialize_field("partial_solutions", &self.partial_solutions)?;
+                combined_puzzle_solution.serialize_field("proof.w", &self.proof.w)?;
+                if let Some(random_v) = &self.proof.random_v {
+                    combined_puzzle_solution.serialize_field("proof.random_v", &random_v)?;
+                }
+                combined_puzzle_solution.end()
+            }
+            false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
+        }
+    }
+}
+
+impl<'de, N: Network> Deserialize<'de> for CoinbaseSolution<N> {
+    /// Deserializes the coinbase solution from a JSON-string or buffer.
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        match deserializer.is_human_readable() {
+            true => {
+                let combined_puzzle_solution = serde_json::Value::deserialize(deserializer)?;
+                Ok(Self::new(
+                    serde_json::from_value(combined_puzzle_solution["partial_solutions"].clone())
```

Comment:
```suggestion
                    serde_json::from_value(combined_puzzle_solution["partial_solutions"].take())
```
([Value::take](https://docs.rs/serde_json/latest/serde_json/enum.Value.html#method.take)) can be used to avoid the clone here and down below

Reply:
I filed https://github.com/AleoHQ/snarkVM/pull/1181 for the existing implementations; these changes are rather trivial, so it's low priority wrt. any conflicts